src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,222 @@
+                    .ContinueWith(task => LogTimeouts(task, x.Key.ToString())));
+        }
 
-                if (_openDocuments.IsEmpty)
-                {
-                    lock (_lock)
-                    {
-                        _queueRunning = false;
-                    }
-                }
-                else
-                {
-                    this.ProcessQueue();
-                }
-            });
+        private Task WaitForInitialStartupWorkIfAny()
+        {
+            return Task.Delay(10 * 1000, _initializationQueueDoneSource.Token)
+                                .ContinueWith(task => LogTimeouts(task, nameof(_initializationQueueDoneSource)));
+        }
+
+        private void LogTimeouts(Task task, string description)
+        {
+            if (!task.IsCanceled) _logger.LogError($"Timeout before work got ready for {description}.");
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs changeEvent)
+        {
+            if (changeEvent.Kind == WorkspaceChangeKind.DocumentChanged
+                || changeEvent.Kind == WorkspaceChangeKind.DocumentAdded
+                || changeEvent.Kind == WorkspaceChangeKind.ProjectAdded)
+            {
+                var project = changeEvent.NewSolution.GetProject(changeEvent.ProjectId);
+                QueueForAnalysis(new[] { project });
+            }
         }
 
-        private async Task Dequeue()
+        private async Task Analyze(Project project, CancellationTokenSource workReadySource, CancellationToken token)
         {
-            var tasks = new List<Task<DiagnosticResult>>();
-            for (var i = 0; i < 50; i++)
+            try
             {
-                if (_openDocuments.IsEmpty)
+                // Only basic syntax check is available if file is miscellanous like orphan .cs file.
+                // Todo: Where this magic string should be moved?
+                if (project.Name == "MiscellaneousFiles")
                 {
-                    break;
+                    await AnalyzeSingleMiscFilesProject(project);
+                    return;
                 }
 
-                if (_openDocuments.TryDequeue(out var filePath))
+                var allAnalyzers = _providers
+                    .SelectMany(x => x.CodeDiagnosticAnalyzerProviders)
+                    .Concat(project.AnalyzerReferences.SelectMany(x => x.GetAnalyzersForAllLanguages()));
+
+                var compiled = await project.WithCompilationOptions(
+                    _rulesetsForProjects.BuildCompilationOptionsWithCurrentRules(project))
+                    .GetCompilationAsync(token);
+
+                ImmutableArray<Diagnostic> results = ImmutableArray<Diagnostic>.Empty;
+
+                if (allAnalyzers.ToImmutableArray().Any())
                 {
-                    tasks.Add(this.ProcessNextItem(filePath));
+                    results = await compiled
```

Comment:
I guess, it might be fine in your case, since you always re-analyze whole project and never analyze same project at the same time.

Reply:
but since it does full re-analysis on a project even for simple char changes, with this approach, squiggle might get updated really slow.

Reply:
well.. might be okay in your case since you seems not doing dependency tracking. meaning changing a char won't cause it to reanalyze all projects that depends on the project current file belong to.

can't say for sure without doing perf analysis. and if it turns out too expensive, we have API that let you to analyze per a file. 

that will require more complex bookkeeping than this implementation. so I would go with simpler approach first.

Reply:
Can you point me right direction with single file api via code or doc? I think current solutions seems to be fast enough but if it can be made faster i think i will write improvements in another PR, any speed to get more responsive user experience is welcome

Reply:
http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/CompilationWithAnalyzers.cs,470

there are multiple of these APIs per different type. all getting tree to show which file.